qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java

```diff
@@ -65,16 +73,38 @@ public void testSnapshotBasedRecovery() throws Exception {
                 break;
             case MIXED:
             case UPGRADED:
-                // the following `if` for first round mixed was added as a selective test mute. Sometimes the primary shard ends
-                // on the upgraded node. This causes issues when removing and adding replicas, since then we cannot allocate to
-                // any of the old nodes. That is an issue only for the first mixed round, hence this check.
-                // Ideally we would find the reason the primary ends on the upgraded node and fix that (or figure out that it
-                // is all good).
-                // @AwaitsFix(bugUrl = https://github.com/elastic/elasticsearch/issues/76595)
-                if (CLUSTER_TYPE != MIXED || FIRST_MIXED_ROUND == false) {
-                    // Drop replicas
-                    updateIndexSettings(indexName, Settings.builder().put(IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 0));
+                if (FIRST_MIXED_ROUND) {
+                    String primaryNodeId = getPrimaryNodeIdOfShard(indexName, 0);
+                    Version primaryNodeVersion = getNodeVersion(primaryNodeId);
+
+                    // Sometimes the primary shard ends on the upgraded node (i.e. after a rebalance)
+                    // This causes issues when removing and adding replicas, since then we cannot allocate to any of the old nodes.
+                    // That is an issue only for the first mixed round.
+                    // In that case we exclude the upgraded node from the shard allocation and cancel the shard to force moving
+                    // the primary to a node in the old version, this allows adding replicas in the first mixed round.
+                    if (primaryNodeVersion.after(UPGRADE_FROM_VERSION)) {
+                        updateIndexSettings(
+                            indexName,
+                            Settings.builder()
+                                .put("index.routing.allocation.exclude._id", primaryNodeId)
+                                .build()
+                        );
+                        cancelShard(indexName, 0, primaryNodeId);
+
+                        String currentPrimaryNodeId = getPrimaryNodeIdOfShard(indexName, 0);
+                        assertThat(getNodeVersion(currentPrimaryNodeId), is(equalTo(UPGRADE_FROM_VERSION)));
+
+                        updateIndexSettings(
```

Comment:
I think this risks balancing back the shard to the upgraded node. You can disable balancing and reenable it after reestablishing replica. Or we could just keep the exclusion until the next round. I think I prefer the latter.